be/src/storage/rowset_update_state.cpp

```diff
@@ -232,20 +232,110 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
+    // _partial_update_states is empty which means write column is empty
+    if (_partial_update_states.empty()) {
+        return Status::InternalError("write column is empty");
+    }
+
+    // _read_version is equal to lastest_applied_version which means there is no other rowset is applied
+    // the data of write_columns can be write to segment file directly
+    if (lastest_applied_version == _read_version) {
+        return Status::OK();
+    }
+
+    LOG(INFO) << "lastest_applied_version: " << lastest_applied_version.to_string()
+              << " vs read_version: " << _read_version.to_string();
+
+    // get rss_rowids to identify conflict exist or not
+    int64_t t_start = MonotonicMillis();
+    uint32_t num_segments = _upserts.size();
+    std::vector<std::vector<uint64_t>> new_rss_rowids;
+    new_rss_rowids.resize(num_segments);
+    for (uint32_t i = 0; i < num_segments; ++i) {
+        auto& pks = *_upserts[i];
+        new_rss_rowids[i].resize(pks.size());
+        index.get(pks, &new_rss_rowids[i]);
+    }
+    int64_t t_get_rowids = MonotonicMillis();
+
+    for (uint32_t i = 0; i < num_segments; ++i) {
+        std::vector<std::unique_ptr<vectorized::Column>> new_write_columns;
+        new_write_columns.resize(_partial_update_states[i].write_columns.size());
+        for (uint32_t j = 0; j < new_write_columns.size(); ++j) {
+            new_write_columns[j] = _partial_update_states[i].write_columns[j]->clone_empty();
+        }
+
+        bool has_conflict = false;
+        uint32_t num_rows = new_rss_rowids[i].size();
+        std::vector<uint32_t> idxes;
+        std::vector<uint32_t> conflict_idxes;
+        std::vector<uint64_t> conflict_rowids;
+        idxes.resize(num_rows);
+        DCHECK_EQ(num_rows, _partial_update_states[i].src_rss_rowids.size());
+        for (size_t j = 0; j < new_rss_rowids[i].size(); ++j) {
+            uint64_t new_rss_rowid = new_rss_rowids[i][j];
+            uint32_t new_rssid = new_rss_rowid >> 32;
+            uint64_t rss_rowid = _partial_update_states[i].src_rss_rowids[j];
+            uint32_t rssid = rss_rowid >> 32;
+
+            if (rssid != new_rssid) {
+                has_conflict = true;
+                idxes[j] = -1;
+                conflict_idxes.emplace_back(j);
+                conflict_rowids.emplace_back(new_rss_rowid);
+            } else {
+                idxes[j] = j;
+            }
+        }
+        if (has_conflict) {
+            size_t num_default = 0;
+            std::map<uint32_t, std::vector<uint32_t>> rowids_by_rssid;
+            std::vector<uint32_t> read_idxes;
+            plan_read_by_rssid(conflict_rowids, &num_default, &rowids_by_rssid, &read_idxes);
+            DCHECK_EQ(conflict_idxes.size(), read_idxes.size());
+            RETURN_IF_ERROR(tablet->updates()->get_column_values(read_column_ids, num_default > 0, rowids_by_rssid,
+                                                                 &_partial_update_states[i].write_columns));
```

Comment:
the number of conflicts is usually very small, could we avoid copying the entire column? just set the rows that need to be updated.